File: 70.climbing-stairs.py
#
# @lc app=leetcode id=70 lang=python3
#
# [70] Climbing Stairs
#

# @lc code=start
class Solution:
    def climbStairs(self, n: int) -> int:
        if n <= 3:
            return n
        oneStep = 1
        twoSteps = 2
        for _ in range(n-2):
            foo = oneStep + twoSteps
            oneStep = [twoSteps][0]
            twoSteps = [foo][0]
        return twoSteps
# @lc code=end

File: 70.climbing-stairs.ts
/*
 * @lc app=leetcode id=70 lang=typescript
 *
 * [70] Climbing Stairs
 */

// @lc code=start
function factorial(x: number): number {
    if (x === 0) return 1;
    return x * factorial(x - 1);
}

function permutation(total: number, a: number, b: number): number {
    return factorial(total) / (factorial(a) * factorial(b));
}

function climbStairs(n: number): number {
    let count = 0;
    let num1steps = n;
    let num2steps = 0;
    for (let numSteps = n; numSteps >= (n / 2); numSteps--) {
        count += permutation(numSteps, num1steps, num2steps);
        num1steps -= 2;
        num2steps++;
    }
    return count;
};
// @lc code=end